# <a href="https://www.pythontutorial.net/advanced-python/python-decimal/" style="color:Tomato">Python Decimal</a>

Trong bài này ta sẽ học về module `decimal` trong Python.

### Tables of Contents
* [Introduction to the Python `decimal` module](#1)
* [Decimal context](#2)
* [Decimal constructor](#3)
* [Decimal arithmetic operations](#4)
* [Summary](#sum)

## <a class="anchor" id="1">Introduction to the Python `decimal` module</a>

Trong Python có những trường hợp <span style="color:DarkOrange">số ở hệ thập phân không thể được biểu diễn chính xác dưới dạng hệ nhị phân</span>. Ví dụ:

In [1]:
x = 0.1
y = 0.1
z = 0.1

s = x + y + z

print(s)

0.30000000000000004


Kết quả không phải là `0.3` mà là `0.30000000000000004`.

Để giải quyết vấn đề này, ta có thể sử dụng class `Decimal` từ module `decimal`:

In [2]:
import decimal
from decimal import Decimal


x = Decimal('0.1')
y = Decimal('0.1')
z = Decimal('0.1')

s = x + y + z

print(s)

0.3


Giờ thì output đã ra như mình mong muốn.

Module `decimal` hỗ trợ các phép toán số học giống như những gì ta được học ở trường.

Không giống như Python, Decimal biểu diễn các số thực một cách chính xác. Kết quả của các phép toán số học cũng được biểu diễn chính xác. Ví dụ, đoạn code sau trả ra kết quả `0.0`:

In [3]:
Decimal('0.1') + Decimal('0.1') + Decimal('0.1') - Decimal('0.3')

Decimal('0.0')

## <a class="anchor" id="2">Decimal context</a>

Decimal luôn gắn với một context kiểm soát hai yếu tố sau:
- Độ chính xác cho các phép tính số học
- Thuật toán làm tròn số

Mặc định, context này sẽ là global. Tuy nhiên ta có thể tự đặt ra một số context chỉ có tác dụng trong local mà không ảnh hưởng đến global.

Để lấy context mặt định của `decimal`, ta sử dụng hàm `getcontext()`:

In [4]:
decimal.getcontext()

Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

Để tạo một context copy từ một context khác, ta sử dụng hàm `localcontext()`:
```python
decimal.localcontext(ctx=None)
```

Hàm `localcontext()` trả về một context mới, được copy từ một context được truyền vào bởi biến `ctx`.

In [5]:
decimal.localcontext(ctx=None)

Khi đã có object context, bạn có thể lấy các thuộc tính độ chính xác và thuật toán làm tròn số của nó như sau:
- `ctx.pre`: get hoặc set độ chính xác. `ctx.pre` là một số kiểu `int`, giá trị mặc định là 28.
- `ctx.rounding`: get hoặc set cơ chế làm tròn số. `ctx.rounding` là một `string`, giá trị mặc định là `"ROUND_HALF_EVEN"`.

Python cung cấp các cơ chế làm tròn số như sau:

| Rounding        | Description                                              |
|-----------------|----------------------------------------------------------|
| ROUND_UP        | round away from zero                                     |
| ROUND_DOWN      | round towards zero                                       |
| ROUND_CEILING   | round to ceiling (towards positive infinity)             |
| ROUND_FLOOR     | round to floor (towards negative infinity)               |
| ROUND_HALF_UP   | round to nearest, ties away from zero                    |
| ROUND_HALF_DOWN | round to nearest, ties towards zero                      |
| ROUND_HALF_EVEN | round to nearest, ties to even (least significant digit) |

Ví dụ sau minh hoạ việc lấy thông tin của default context:

In [6]:
import decimal

ctx = decimal.getcontext()

print(ctx.prec)
print(ctx.rounding)

28
ROUND_HALF_EVEN


Ví dụ sau minh hoạ về cách hoạt động của cơ chế làm tròn số `"ROUND_HALF_EVEN"`:

In [7]:
import decimal
from decimal import Decimal


x = Decimal('2.25')
y = Decimal('3.35')

print(round(x, 1))
print(round(y, 1))

2.2
3.4


Thay đổi cơ chế làm tròn số thành `"ROUND_HALF_UP"`, ta sẽ được kết quả khác:

In [8]:
import decimal
from decimal import Decimal


ctx = decimal.getcontext()
ctx.rounding = decimal.ROUND_HALF_UP

x = Decimal('2.25')
y = Decimal('3.35')

print(round(x, 1))
print(round(y, 1))

2.3
3.4


Ví dụ sau copy context mặc định và đổi cơ chế làm tròn số thành `"ROUND_HALF_UP"`:

In [9]:
import decimal
from decimal import Decimal


x = Decimal('2.25')
y = Decimal('3.35')

with decimal.localcontext() as ctx:
    print('Local context:')
    ctx.rounding = decimal.ROUND_HALF_UP
    print(round(x, 1))
    print(round(y, 1))

print('Global context:')
print(round(x, 1))
print(round(y, 1))

Local context:
2.3
3.4
Global context:
2.3
3.4


Lưu ý là local context không ảnh hưởng đến global context. Local context chỉ có ảnh hưởng trong block `with`.

## <a class="anchor" id="3">Decimal constructor</a>

`Decimal` constructor dùng để tạo một `Decimal` object từ một giá trị cho trước:

In [10]:
Decimal(value='0', context=None)

Decimal('0')

Tham số `value` có thể là một integer, string, tuple, float, hoặc một `Decimal` object khác. Nếu không truyền giá trị `value` thì nó nhận giá trị mặc định là `0`.

Nếu `value` là một tuple, nó phải có dạng như sau:
```python
(sign, (digit1,digit2, digit3,...), exponent)
```
trong đó:
- `sign` là dấu: `0` là dương còn `1` là âm
- `(digit1,digit2, digit3,...)` là một tuple: các chữ số
- `exponent`: số mũ

Ví dụ: Ta muốn biểu diễn số `3.14`:

In [11]:
import decimal
from decimal import Decimal

x = Decimal((0, (3, 1, 4), -2))
print(x)

3.14


Giải thích: $3.14 = 314 \times 10^{-2}$

Lưu ý: độ chính xác của context không ảnh hưởng đến constructor. Ví dụ:

In [12]:
import decimal
from decimal import Decimal


decimal.getcontext().prec = 2

pi = Decimal('3.14159')
radius = 1

print(pi)

area = pi * radius * radius
print(area)

3.14159
3.1


Khi dùng một số thực không thể được biểu diễn chính xác trong hệ nhị phân, `Decimal` constructor cũng không thể biểu diễn nó một cách chính xác:

In [13]:
import decimal
from decimal import Decimal

x = Decimal(0.1)
print(x)

0.1000000000000000055511151231257827021181583404541015625


Trong thực tế, ta <span style="color:DarkOrange">nên sử dụng kiểu string hoặc tuple cho tham số của `Decimal` constructor</span>.

## <a class="anchor" id="4">Decimal arithmetic operations</a>

Các toán tử số học của `Decimal` sẽ hơi khác. Ví dụ phép div:
```python
x // y = trunc( x / y)
```

Không phải phép toán nào của module `math` cũng được hỗ trợ trong `Decimal`.

Khi sử dụng các phép toán của module `math` cho các `Decimal` object, `math` sẽ cast `Decimal` object thành float trước khi thực hiện chúng. Nó sẽ khiến cho sự chính xác của `Decimal` object không còn.

## <a class="anchor" id="sum" style="color:Violet"> Tổng kết </a>

- Python sử dụng module `decimal` để biểu diễn chính xác số thực.
- Sử dụng class `Decimal` từ module `decimal` để tạo ra các `Decimal` object từ strings, integers, tuples.
- Các số `Decimal` có một context để điều khiển độ chính xác và thuật toán làm tròn.
- Class `Decimal` không hỗ trợ tất cả các phép toán của module `math`. Nếu có thể, ta nên sử dụng các phép toán số học của `Decimal`.